packages/web-components/src/utils/element-internals.ts

```diff
@@ -60,3 +64,46 @@ export function toggleState(elementInternals: ElementInternals, state: string, f
   // @ts-expect-error - Baseline 2024
   elementInternals.states.delete(state);
 }
+
+/**
+ * Check if a given attribute value is a valid state. Attribute values are often kebab-cased, so this function converts
+ * the kebab-cased `state` to camelCase and checks if it exists in as a key in the `States` object.
+ *
+ * @param States  - the object containing valid states for the attribute
+ * @param state - the state to check
+ * @returns true if the state is in the States object
+ * @internal
+ */
+export function hasMatchingState(States: Record<string, string> | undefined, state: string | undefined): boolean {
+  if (!States || !state) {
+    return false;
+  }
+
+  const camelState = state.replace(/-([a-z])/g, g => g[1].toUpperCase());
```

Comment:
I don't think this is a blocking issue, but I do feel this is making some assumptions about how components are built while we don't have any tooling to enforce this. Would something like this more robust: `return Object.values(States).includes(state)`?

Reply:
Checking `Object.values(States).includes(state)` is less performant than using `hasOwn`, but I agree that assuming keys will be formatted correctly isn't a great approach.

Reply:
This is now updated to use a `WeakMap`, which uses the `States` object as the key and a `Set` to store `Object.values(States)` as the value.

Reply:
This is nice!